### ISSUE TITLE
End user needs more direct low level control over pango markup of prefix/suffix icons

### ISSUE THREAD

**Issue by somospocos** [tobi-wan-kenobi/bumblebee-status#531]
I am "guilty" of lobbying for pango markup support (issue #493) and custom prefix/suffix colors (issue #520) and I've just bumped into a situation which proves that the quick hacks bumblebee has got for making me happy, might not be enough in the long term.

I took a good beating from the technical debt, so hopefully this issue will influence the ongoing redesign of bumblebee in the development branch.

While bumblebee supports FontAwesome for icons, this might not be enough for a capricious and picky user (like me) for whatever reasons, technical or aesthetical ones.

I have been toying with Nerd patched fonts (https://www.nerdfonts.com/) which bundle FontAwesome icons and some other icon packs. I did so by making a copy of **awesome-fonts.json**, renaming it to **nerd-fonts.json** and changing things there, then using **-i nerd-fonts** when invoking bumblebee-status.

I have been experimenting with a narrow font (Iosevka) and a smallish font size, for the sake of saving screen space and packing as much info as possible in a single status line. I have noticed that for narrow fonts, the icons are **very** small, way smaller than providing a good visual experience.

I was able to come up with a workaround, but... not in context of bumblebee-status.

This is what I've been playing with:

The Nerd patched font Iosevka that can be downloaded from here https://github.com/ryanoasis/nerd-fonts/releases/download/v2.0.0/Iosevka.zip

    bar {
        id top
        workspace_buttons no
        font pango:Iosevka 9
        position top
        status_command test.sh
    }

this is the relevant snippet from the i3 config file.

And this is the status command script:

    #!/bin/sh

    echo '{"version": 1}'
    echo '['
    echo '[]'

    while true; do
        echo ',[{"full_text": "tiny icon", "markup": "pango"}, {"full_text": "", "markup": "pango"}, {"full_text": "some text", "markup": "pango"}, {"full_text": "adjusted size icon", "markup": "pango"}, {"full_text": "<span font_desc=\"Iosevka 16\"></span>", "markup": "pango"}, {"full_text": "some text", "markup": "pango"}, {"full_text": "adjusted size + adjusted offset icon", "markup": "pango"}, {"full_text": "<span font_desc=\"Iosevka 16\" rise=\"7500\"></span>", "markup": "pango"}, {"full_text": "some text", "markup": "pango"}]'
        sleep 1
    done

You can see that the default icon is super tiny. By adjusting font size using pango markup, I got a better result. By further adjusting a bit the vertical offset, I got an even better result.

The question is: "how to do the same with bumblebee?" So far I don't know an answer. To hardcode the pango markup for every icon in the icons json file? No, because icons are font size agnostic and can be used by many themes. To have a copy of the icons file with needed pango markup just for my particular theme and bumblebee-status config? Maybe... oh wait, it's a bad idea. If one would use custom colors for prefixes/suffixes in a theme, then, according to #520, bumblebee will already generate pango markup internally for that.

Which reminds me about this: https://github.com/tobi-wan-kenobi/bumblebee-status/issues/520#issuecomment-573314665

Perhaps bumblebee should keep supporting the settings added in the aforementioned #520, but also support a whole dictionary of settings for prefix/suffix, just like the comment above suggests? And by whole I mean all possible attributes that are specified here https://developer.gnome.org/pango/stable/pango-Markup.html . But where exactly? Hardcoding font size values in themes feels a bit rough too, since themes are supposed to be entirely font-agnostic, right?

This is very far from being a feature request, for now it's rather a reference request for comments.

**Comment by tobi-wan-kenobi** [comment]
Excellent points @somospocos , and in fact, the whole pango markup discussion was what sparked my ideas of a re-write. Clearly, the engine doesn't offer the required flexibility at the moment.

One thing that is clear to me is that I am unwilling to expose any kind of additional complexity to the user. Having said that, the fact that the font size is configured outside the context of bumblebee, and - as far as I know - bumblebee doesn't have access to the font size configured for the bar, makes things *very* interesting. Maybe using relative sizes (smaller, larger) might help, but then, that's pretty restricted.

If possible, I'd also like to hide the fact that pango is being used, at least for the "average" user. Maybe a config parameter such as "icon size" might be sufficient.

And yes, you are 100% right, themes should be font agnostic.

Ideally, such parts of the pango markup are something that is "mixed in" separately into the icon set, so hat it can be used with arbitrary icon sets. The sets themselves, of course, would also be able to use pango attributes (as you said, generically, ideally).

@somospocos many thanks for the well-reasoned and formulated arguments, this post will definitely find its way into the redesign's philosophy.

Thanks!

**Comment by somospocos** [comment]
> If possible, I'd also like to hide the fact that pango is being used, at least for the "average" user. Maybe a config parameter such as "icon size" might be sufficient.

You can see above that I had to modify at least two parameters to achieve the desired result: the font size and that "rise" thing which I learned about very recently, while trying to solve the problem, by reading the pango syntax documentation. This means that the config parameter approach would require at least two parameters for icons. And then my developing paranoia about the whole issue whispers "what if some other future scenario would require modifying more than two parameters?". That's why I've been thinking about a way to somehow expose the full access to the pango markup to the user, instead of config parameters, rather something like a whole template for what will become the pango markup for the icon in the output of bumblebee that gets passed to i3bar. By analogy with **.format** parameters that certain modules have, maybe there could be a global icon parameter like

    icon.markuptemplate="<span font_desc=\"Iosevka 16\" rise=\"7500\">{}</span>"

This would allow the user to hack the pango markup as much as he wants and internally, bumblebee would just insert the unicode icon char instead of the **{}** thing while doing string formatting.

Of course this would be a fully optional thing with default value **"{}"** (no pango markup at all).

**Comment by somospocos** [comment]
PS: such a markup template for icons would be a more general solution to having only custom icon fg/bg colors, so it could entirely replace (and greatly enhance!) #520. And existing themes wouldn't need to be modified at all: bumblebee could use settings from themes and once it spots a **.startswith("&lt;span")** pattern in the icon markup template, it could use the template instead. The only possible problem that I see here is that an explicit pango markup template would need a way to inherit the fg/bg color from the theme if it doesn't specify one itself (for themes that have "cycle" blocks).

**Comment by tobi-wan-kenobi** [comment]
Currently, I'm thinking something like adding a `markup: { <attribute>: <value> }` capability to the themes, and add an easy way to override generic attributes (e.g. `-p theme.prefix.markup.size=12`).

The parameters would be a bit lengthy, but the advantage would be that it's markup-type agnostic (i.e. also works if we ever have markup other than pango), and avoids all "merging" issues (i.e. the theme provides something, but the user can override it), and the user wouldn't have to know too much about the implementation details.

And, yes, that would mean that in your specific use-case, you'd have to set two parameters, but you do want to change two values of the markup, so that seems OK (`-p theme.prefix.markup.font_desc="Iosevka 16" theme.prefix.markup.rise="7500"`).

If that seems too heavy, maybe an advanced user can specify a stringified JSON that'll be merged into the JSON.

How does that sound?

**Comment by somospocos** [comment]
> And, yes, that would mean that in your specific use-case, you'd have to set two parameters, but you do want to change two values of the markup, so that seems OK (-p theme.prefix.markup.font_desc="Iosevka 16" theme.prefix.markup.rise="7500").

My biggest concern with this approach is that to me it feels like it doesn't scale very well. Imagine the exaggerate hypothetical situation that an user may want to customize **all** pango markup attributes. There are approx 30 attributes (a few of them overlapping). This would lead to a giant length of the bumblebee-status invocation command, because it would contain **theme.prefix.markup.** 30 times. In this case as an end user I would certainly prefer having just one parameter with the markup template, as described above. After all, the user still needs to consult the pango markup docs to learn the correct attribute names. And not needing to provide a bunch of parameters with same prefix string in their name would just save a lot of typing.

I also expect that troubles happening with the multiple parameters approach might end being a bit harder to debug. Recalling the incident that happened to me some time ago when I mistyped a parameter name, for debugging such an issue there would be two options:

1. either have validation of bumblebee parameter names, so they all match pango markup attribute names, which would increase the amount of code that would have to be written

2. run bumblebee-status in console (or with debug mode on) and visually inspect the output it produces to identify the problem

I wonder if the "Talk is cheap, show me the code" famous quote by Linus applies here and we should try coding some minimal experimental prototypes with both approaches, then compare them.

PS: I also wonder if it's possible to customize a pango markup string with something similar to CSS for HTML markup. I did a few searches about it, but didn't find anything yet. And it's very likely that if such a thing exists, it may require bumblebee-status to get a big dependency, maybe something like a whole Python binding for GTK, which isn't good.

**Comment by somospocos** [comment]
After giving it some more thought, I'm really tempted to give it a try implementing an **iconmarkuptemplate** global parameter as described above in a separate branch, fully backwards compatible, thus on top of all already existing functionality, including custom prefix/suffix color settings in themes (instead of replacing that feature). Hypothetically, if I ever get there (albeit I feel like this would require slightly more effort than any of my previous contributions), I will for sure use that branch for my personal needs (after all, this is the whole point of thinking to do it). If you'll want the change in trunk - it will be even better, since I wouldn't need to spend additional time keeping that separate branch in sync with master.

However, I see a show stopper for the whole thing above, which requires an intermediate step. **bumblebee.output.I3BarOutput.draw()** will get even bigger: 

- it already takes two screens in terms of lines of code
- it will have yet another level of complexity for prefixes - three layers:
  1. prefixes getting colors from theme by default
  2. prefixes getting custom colors from theme (#520)
  3. prefixes getting full customization from custom pango markup format, while inheriting existing settings from 1. and 2. if same settings are not explicitly present in the custom pango format string (this is what I'm tempted to implement)

At this point I would really like having this draw() function extracted into its own class, something like **WidgetDrawer** (since it accepts a widget as argument) and split into smaller methods.

Do you mind if I attempt such a refactoring?

**Comment by tobi-wan-kenobi** [comment]
I don't mind at all, on the contrary. I think it sounds like a good suggestion and I will gladly review the PR, once it's there.

And from your contributions so far, I am convinced the changes will improve bumblebee and be a worthwhile addition, you are an extremely capable developer!

⁣

**Comment by somospocos** [comment]
@tobi-wan-kenobi please have a look at

https://github.com/somospocos/bumblebee-status/tree/refactor-draw

no pull request yet, I'm still in doubt myself if this is a code improvement or a code mutilation. The only goal was, as I said above, to attempt splitting the big **draw()** function into smaller parts.

**Comment by tobi-wan-kenobi** [comment]
@somospocos sorry for taking so long to review!

The changes look good to me, and I think they are an improvement over what is there currently.

Just 2 remarks:

1. You've probably seen the development branch - stuff will change significantly there, and having more modular output is one of the goals. So, fair warning: I'll likely draw a lot of inspiration from what you're doing, but chances are, a lot of stuff will be discarded as well. Can't say when develop will be ready to merge back, though (might be years even, given the time I can spend on the project right now)

2. Some of the things (prefix colors, etc.) I guess should even be moved out of the i3 output class and probably belong in the Theme class, maybe? I think part of the problem was my attempt to over-generalize the concept, and make theme output-agnostic (which is a joke, really - bumblebee-status will never support anything else than i3 anyhow).

But if you sent a PR looking like that, I'd probably just merge it after some rough testing :)

Thanks!

**Comment by somospocos** [comment]
> development branch - stuff will change significantly there

As I said before, I'm totally fine with that. Even if next-gen bumblebee-status (what's happening in the development branch) will have some backwards incompatibilities, I'll always have the choice of either sticking to the then-legacy version (current master branch now) or bug you again with pull requests :D

> I'll likely draw a lot of inspiration from what you're doing, but chances are, a lot of stuff will be discarded as well.

Again, as I said before, I'll be fine with anything (in the technical sense) that is able to solve my particular problems and caprices.

> Can't say when develop will be ready to merge back

At this point I'm not even thinking much about it, I treat everything what's happening in this issue in context of current master branch.

> Some of the things (prefix colors, etc.) I guess should even be moved out of the i3 output class and probably belong in the Theme class, maybe?

So far I am not sure about it. For example the hack that I really want to implement next (the one that is supposed to solve my "too-small-icons compared to the rest of text" trouble) is going to fix the very particular problem of my setup (choice of font name / size) rather than a limitation of bumblebee-status theming mechanism. But we can go into more details on this later, if I succeed to implement what I have in mind.

> bumblebee-status will never support anything else than i3 anyhow

I found this funny, because I have to admit that at some point I wondered about this. Personally, I wouldn't mind to be able to configure my also beloved dzen2 via bumblebee-status, because I like the approach, but this is clearly something that shouldn't be taken into account for now. Maybe later in the far future, when development branch will be finished and merged into master.

Back on topic: despite master having moved some commits ahead, fortunately the refactoring branch still reports as able to merge (saving me from the hassle of doing rebases and resolving conflicts - git wizardry that I never did before), so I'll create a pull request in a few moments.

**Comment by tobi-wan-kenobi** [comment]
@somospocos Just fixed this issue - hopefully for good! - in the development branch.

Basically, what I have added is the possibility to specify *arbitrary* pango attributes inside the theme, in a special "pango" section.

This has a couple of "side effects":
- When a pango section is encountered, that particular widget is automatically written in pango - no need to have a switch on the CLI for pango support
- When a pango section is present, it overrides whatever non-pango formatting the theme would be specifying (ie. for a given widget, it's always either pango or "non-pango", you cannot mix)
- For now, a module writer cannot use pango directly, though I have some ideas on how to change that.

I've also committed a practical example in powerline-pango to demonstrate.

Does that fit your requirements?

The only potential issue, remembering your specific requirements, is that it currently is not possible to change the size of a single separator, but I *think* I can extend the system to incorporate that, as well.

Any thoughts/feedback would be highly appreciated!

(and yes, the implementation in core/output.py is still rough, but in general, I like how theme.py just gets the attributes and output.py has to care about assembling it - though putting that in a util.pango library might be even nicer)

**Comment by somospocos** [comment]
> I've also committed a practical example in powerline-pango to demonstrate.

By looking at it, I don't understand how would I set prefixfg. In the master branch, in themes, "prefixfg" is a key of "defaults" dictionary.

> currently is not possible to change the size of a single separator

I didn't do that to separators, I did it to prefix icons. This is an example I'm using:

    --iconmarkup="<span font_desc='Iosevka 15' rise='-3500'>{}</span>"

How is this going to look inside the theme? Something like

    "defaults": {
        "icons": {
            "pango": {
                "foreground": "#ff0000",
                "font_desc": "Iosevka 15",
                "rise": -3500
            }
        }
    }

?

Here "foreground" is prefixfg from above.

**Comment by tobi-wan-kenobi** [comment]
Ah, yes, I hadn't considered that.

I'd go for something like this:

Such parameters (prefix, suffix, separator) can be either scalar (i.e. the actual value) or dicts, in which case they simply contain a "pango" subsection as usual, probably with a special field "full_text" or so that gives the actual content.

Haven't had time to try that yet, but should work.

And yeah, adding a generic defaults pango handling for prefixes etc., as you've shown, would then also be required. That might be a bit more involved.

Or I go the simple way and say, if the value is scalar, but there's a pango "upstream", use the pango for formatting, but the scalar as value. I guess that should do the trick, as well. But depends on actually trying out. Hopefully will have something tomorrow, I have a 1-2h slot for coding there :P

**Comment by tobi-wan-kenobi** [comment]
I've updated the code in development, the output module is much nicer now, but the theme still needs improvement.

Anyhow, defaults should now work exactly as you showed. I tested by setting all module prefixes to red with a single pango line in the theme's "default" section.

**Comment by somospocos** [comment]
I just did a quick test and I confirm that my use case works in development branch.

This is what I did:

fetched latest changes from development branch of your repo into my local copy, switched to development branch locally and merged it with upstream

used a simple invocation:

    bumblebee-status -m cpu -t powerline-pango-custom

created the powerline-pango-custom.json theme file in ~/.config/bumblebee-status/themes with simplified content:

    {
        "icons": [ "awesome-fonts" ],
        "defaults": {
            "prefix": {
                "pango": {
                    "foreground": "#ff0000",
                    "font_desc": "Iosevka 15",
                    "rise": -3500
                }
            }
        }
    }
    
I've got the expected result - the prefix icon is red and its size/position is adjusted. Feel free to use this simple snippet as manual test case and add it to the docs as example.
